### PR TITLE
feat: relationship between kubernetes secret and other resources

### LINKED ISSUE #9856: [Models]: Relationships for Kubernetes Secrets
### Current Situation
Meshery has only a few relationships for handling Kubernetes Secrets.

### Proposed Change
1. Identify all possible relationships.
1. Propose the relationship `kind` and `type` (see instructions below).
1. Define the constraints for that relationship.

See https://github.com/meshery/meshery/pull/9816 as an example.

---

### Contributor [Guides](https://docs.meshery.io/project/contributing) and [Handbook](https://layer5.io/community/handbook)

- [MeshModel Repository](https://github.com/meshery/meshery/tree/master/server/meshmodel)
- 🖥 [MeshModel Overview](https://docs.google.com/presentation/d/1SQMfyu5shjpGKlYONdVzOtd7UYTgLWBcgUvHMLCZ2tY/edit#slide=id.g226f5de5180_19_259)
- 🙋🏾🙋🏼 Questions: [Discussion Forum](http://discuss.meshery.io) and [Community Slack](https://slack.meshery.io)

## Contributing to Meshery Models

!! See https://docs.meshery.io/project/contributing/contributing-models

### Instructions for Models

_Forthcoming_

### Instructions for Components
While the default shape for new components is a circle, each component should be considered for its best-fit shape.
1. Review and familiarize with the available set of predefined relationship types. Refer the Cytoscape [node types](https://js.cytoscape.org/demos/node-types/) for a list of possible shapes. 
1. Propose a specific shape, best-suited to visually represent the Component. _Example - Deployment as a pentagon._
1. Proposee a specific icon, best-suited to visually represent the Component. _Example - DaemonSet as a skull icon._

### Instructions for Relationships
1. Identify the relationship and any specific constraints to be enforced between the two specific components, their models or potentially other components, models, or environmental considerations.
1. Propose a specific visual representation for the relationship. Visual representation examples: [Hierarchical](https://docs.meshery.io/assets/img/meshmodel/relationships/hierarchical_relationship.png), [Sibling](https://docs.meshery.io/assets/img/meshmodel/relationships/sibling_relationship.png), [Binding](https://docs.meshery.io/assets/img/meshmodel/relationships/binding_relationship.png), [Edge](https://docs.meshery.io/assets/img/meshmodel/relationships/mount_edge_relationship.png)
1. Prospose the appropriate relationship type, using one of the predefined set of relationship types or suggest a new relationship where an existing type does not fit.
1. Create a Relationship Definition (yaml). See the [Relationship Schema](https://github.com/meshery/meshery/tree/master/server/meshmodel/schemas) in MeshModel and [examples]([url](https://github.com/meshery/meshery/tree/master/server/meshmodel/relationships)).
1. Create a policy for evaluation of the relationship (rego). See examples.

### Instructions for Policies
1. _Forthcoming_
### Instructions for Policies
1. _Forthcoming_

### PR DESCRIPTION
**Pull request by RipulHandoo**
**Notes for Reviewers**

This PR fixes #9856 

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Review by leecalcote — APPROVED** [review]
Thanks, @RipulHandoo. Did you update the Meshery Integrations spreadsheet with this new relationship?

File: server/meshmodel/kubernetes/relationships/hierarchical_secret.json [hunk 1]
```diff
@@ -0,0 +1,172 @@
+{
+  "apiVersion": "core.meshery.io/v1alpha1",
+  "kind": "Hierarchical",
+  "metadata": {
+      "description": "A relationship that assigns configuration data for resources deployed in a cluster by environment variable.",
```
Comment (leecalcote) [inline server/meshmodel/kubernetes/relationships/hierarchical_secret.json:5]: @RipulHandoo isn't this the description for ConfigMap? Any changes for Secret?

Comment (leecalcote) [reply server/meshmodel/kubernetes/relationships/hierarchical_secret.json:5]:
```suggestion
      "description": "Use this relationship to make sensitive information available to other resources, like your application, by environment variable.",
```

File: server/meshmodel/kubernetes/relationships/hierarchical_secret.json [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  "model": {
+      "name": "kubernetes",
+      "version": "v1.25.2",
+      "category": {
+          "name": "Orchestration \\u0026 Management",
+          "metadata": null
+      },
+      "metadata": {}
+  },
+  "subType": "Secret",
```
Comment (leecalcote) [inline server/meshmodel/kubernetes/relationships/hierarchical_secret.json:17]: @MUzairS15, this need to be `Secret` temporarily until you've added `SelectionSets` (or arrays of selections), right?

File: server/meshmodel/kubernetes/relationships/hierarchical_secret.json (hunk 1 above)
Comment (RipulHandoo) [reply server/meshmodel/kubernetes/relationships/hierarchical_secret.json]: Done.

Comment (RipulHandoo) [reply server/meshmodel/kubernetes/relationships/hierarchical_secret.json]: Done

**Comment by saurabh100ni** [comment]
Hi @RipulHandoo  ,
Thanks for your PR! I encourage you to share this in the Meshery Development Meeting tomorrow at 8:30 PM IST. Share your screen, and walk us through it. It's cool to take and share feedback. Check the [Doc](https://docs.google.com/document/d/1Dg4Uz9rGTuGHmYZwIt47z1j5x3LmnpxD1M1SgypAkQ0/edit) here to include this PR.

You're welcome at the meeting

Check to [community calendar](https://layer5.io/community/calendar) for more details of meetings at Layers.

**Comment by leecalcote** [comment]
This PR is blocked by and dependent upon changes to be done by, @MUzairS15.

**Comment by RipulHandoo** [comment]
@MUzairS15 @leecalcote This PR is ready to merge. I've thoroughly tested the changes, and the relationship functionality is working as expected.
https://github.com/meshery/meshery/assets/107461226/7a75f82a-dbd4-4073-b851-2e2ac6fb2ee2

File: server/meshmodel/kubernetes/relationships/hierarchical_inventory.json [hunk 3]
```diff
@@ -381,6 +381,159 @@
...
+                "kind": "Secret",
+                "model": "kubernetes",
+                "patch": {
+                    "patchStrategy": "replace",
+                    "mutatorRef": [
+                        [
+                            "name"
+                        ]
+                    ],
+                    "description": "In Kubernetes, Secrets are versatile resources that can be referenced by various other resources to provide configuration data to applications or other Kubernetes resources.\n\nBy referencing Secrets in these various contexts, you can centralize and manage configuration data more efficiently, allowing for easier updates, versioning, and maintenance of configurations in a Kubernetes environment."
```
Comment (MUzairS15) [inline server/meshmodel/kubernetes/relationships/hierarchical_inventory.json:398]:
```suggestion
                    "description": "In Kubernetes, Secrets are versatile resources that can be referenced by various other resources to provide configuration data to applications or other Kubernetes resources securely.\n\nBy referencing Secrets in these various contexts, you can centralize and manage configuration data more efficiently, allowing for easier updates, versioning, and maintenance of configurations in a Kubernetes environment."
```

File: server/meshmodel/kubernetes/relationships/hierarchical_inventory.json [hunk 4]
```diff
@@ -381,6 +381,159 @@
...
+                        [
+                            "spec",
+                            "containers",
+                            "_",
+                            "envFrom",
+                            "secretRef",
+                            "name"
+                        ]
+                    ],
+                    "description": "When deploying applications using ReplicaSets, you may need to configure each replica with specific settings, such as environment variables or configuration files. Secrets can be referenced in the Pod template specification of a ReplicaSet to provide configuration data to each replica pod. DaemonSets, which ensure that a copy of a pod runs on each node, can utilize Secrets to supply configuration data to their pods.\n\nThe keys from the Secret will be exposed as environment variables to the containers within the pods managed by the ReplicaSet."
```
Comment (MUzairS15) [inline server/meshmodel/kubernetes/relationships/hierarchical_inventory.json:472]:
```suggestion
                    "description": "When deploying applications using ReplicaSets, you may need to configure each replica with specific settings, such as environment variables or configuration files. Secrets can be referenced in the Pod template specification of a ReplicaSet to provide configuration data to each replica pod."
```

**Comment by MUzairS15** [comment]
`mutatedRef` path is incorrect.

File: server/meshmodel/kubernetes/relationships/hierarchical_inventory.json [hunk 5]
```diff
@@ -381,6 +381,159 @@
...
+                }
+            },
+            {
+                "kind": "ReplicaSet",
+                "model": "kubernetes",
+                "patch": {
+                    "patchStrategy": "replace",
+                    "mutatedRef": [
+                        [
+                            "spec",
```
Comment (MUzairS15) [inline server/meshmodel/kubernetes/relationships/hierarchical_inventory.json:464]:
```suggestion
                            "spec",
                            "template",
```

File: server/meshmodel/kubernetes/relationships/hierarchical_inventory.json [hunk 6]
```diff
@@ -381,6 +381,159 @@
...
+                }
+              },
+            {
+                "kind": "DaemonSet",
+                "model": "kubernetes",
+                "patch": {
+                    "patchStrategy": "replace",
+                    "mutatedRef": [
+                        [
+                            "spec",
```
Comment (MUzairS15) [inline server/meshmodel/kubernetes/relationships/hierarchical_inventory.json:446]:
```suggestion
                            "spec",
                            "template"
```

File: server/meshmodel/kubernetes/relationships/hierarchical_inventory.json [hunk 7]
```diff
@@ -381,6 +381,161 @@
...
+        ],
+        "to": [
+            {
+                "kind": "Deployment",
+                "model": "kubernetes",
+                "patch": {
+                    "patchStrategy": "replace",
+                    "mutatedRef": [
+                        [
+                            "spec",
```
Comment (MUzairS15) [inline server/meshmodel/kubernetes/relationships/hierarchical_inventory.json:410]:
```suggestion
                            "spec",
                            "template",
```

**Review by leecalcote — DISMISSED** [review]

**Review by MUzairS15 — COMMENTED** [review]
4 new relationships are to be added as soon as minor updates that are required are committed.
Nice work @RipulHandoo

**Comment by RipulHandoo** [comment]
@MUzairS15 This PR is ready to merge and Working as expected.

https://github.com/meshery/meshery/assets/107461226/3b112123-d7fc-4c4f-ae82-2b73dc46916a

**Comment by leecalcote** [comment]
Super cool, @RipulHandoo 🙂

**Pull request merged**